src/main/java/com/jcabi/s3/AwsListIterator.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Iterator for large lists returned by S3.
+ * @author Roman Kisilenko (roman.kisilenko@gmail.com)
+ * @version $Id$
+ * @since 0.10
+ */
+class AwsListIterator implements Iterator<String> {
+
+    /**
+     * Region we're in.
+     */
+    private final transient Region region;
+
+    /**
+     * Bucket name.
+     */
+    private final transient String bucket;
+
+    /**
+     * Prefix to use.
+     */
+    private final transient String prefix;
+
+    /**
+     * Partial S3 response iterator.
+     */
+    private transient Iterator<String> partial;
+
+    /**
+     * Next marker for partial response or null if response contains no more
+     * data.
+     */
+    private transient String marker;
+
+    /**
+     * Constructs AwsListIterator.
+     * @param pfx Key prefix
+     * @param rgn Region we're in
+     * @param bkt Bucket name
+     */
+    public AwsListIterator(final Region rgn, final String bkt,
+        final String pfx) {
+        this.prefix = pfx;
+        this.region = rgn;
+        this.bucket = bkt;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final boolean hasNext() {
+        this.loadDataIfNeeded();
+        return this.partial.hasNext();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final String next() {
+        if (!this.hasNext()) {
+            throw new NoSuchElementException(
+                "There are no more elements in this iterator"
+            );
+        }
+        return this.partial.next();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void remove() {
+        throw new UnsupportedOperationException("Remove is not supported");
+    }
+
+    /**
+     * Loads next portion of data from S3 if current data is already
+     * consumed and more data is available.
+     */
+    private void loadDataIfNeeded() {
+        if (this.partial == null || (!this.partial.hasNext()
+            && this.marker != null)) {
+            this.loadPartialData();
+        }
+    }
+
+    /**
+     * Loads next portion of data from S3.
+     */
+    private void loadPartialData() {
+        try {
+            final AmazonS3 aws = this.region.aws();
+            final long start = System.currentTimeMillis();
+            final ObjectListing listing = aws.listObjects(
+                new ListObjectsRequest()
+                    .withBucketName(this.bucket)
+                    .withPrefix(this.prefix)
+                    .withMarker(this.marker)
+            );
+            this.marker = listing.getNextMarker();
+            final Collection<String> list = new LinkedList<String>();
+            for (final S3ObjectSummary sum
+                : listing.getObjectSummaries()) {
+                list.add(sum.getKey());
+            }
+            this.partial = list.iterator();
+            Logger.info(
+                this,
+                StringUtils.join(
```

Comment:
better put it in one line and add `// @checkstyle LineLength (1 line)` before it